### PR TITLE
RM3 refactoring to eliminate custom stopwords lists

### PR DESCRIPTION
**Pull request by lintool**
tl;dr -

+ Got rid of hard-coded stopwords list, replaced with _df_ threshold
+ Set the _df_ threshold heuristically without any tuning
+ Effectiveness on newswire went up; effectiveness on everything else stayed about the same (minor differences)

Getting rid of RM3 stopwords: I have long known that stopwords have a big impact in tuning RM3.
If we include stopwords in feedback, effectiveness is affected negatively. A natural way to build a stopwords list is by selecting top _k_ words from the collection, which is the existing implementation. However, we only had two stopwords lists, for gov2 and for Twitter. The gov2 list is used on all collections other than Twitter.

This PR gets rid of the custom stopwords lists and replaces them with a _df_ threshold: If a term appears in more than _n_ percent of the documents, then it is discarded as a feedback term.

This heuristic has the advantage of getting rid of the stopword lists, which need to be collection specific, but at the cost of introducing an additional tuning parameter.

Cognizant of the dangers of (essentially) tuning on test data, here's what I did:

+ For newswire collections, I picked a number, 10%, that seemed right. This value actually increased effectiveness in most conditions across all newswire collections.
+ This 10% value worked fine on web collections also; basically, effectiveness didn't change.

Since this was the first and only heuristic value I selected, it's *not* considered parameter tuning.

The 10% threshold, however, doesn't work well on tweets, because the "documents" are much shorter.
I used the `ExtractRm3Stopwords` util to list terms in the collection by _df_: For the Tweets2011 collection, I found a threshold close to a nice round number that approximated the length of the current stopwords list, by eyeballing the _df_ values. This turned out to be 1%. I did this again for the Tweets2013 collection, using the same approach, and obtained a value of 0.7%.

With both values, we obtained effectiveness pretty close to the old values with the hard-coded stopwords list.

Detailed changes in effectiveness:

```
# Disks 1 + 2
Topics  51-100: bm25+rm3 : map : 0.2582 0.2612  !
Topics 101-150: bm25+rm3 : map : 0.2537 0.2582  !
Topics 151-200: bm25+rm3 : map : 0.3136 0.3214  !
Topics  51-100: ql+rm3   : map : 0.2475 0.2502  !
Topics 101-150: ql+rm3   : map : 0.2437 0.2477  !
Topics 151-200: ql+rm3   : map : 0.2940 0.3015  !
Topics  51-100: bm25+rm3 : p30 : 0.4800 0.4853  !
Topics 101-150: bm25+rm3 : p30 : 0.4560 0.4573  !
Topics 151-200: bm25+rm3 : p30 : 0.5067 0.5087  !
Topics  51-100: ql+rm3   : p30 : 0.4633 0.4700  !
Topics 101-150: ql+rm3   : p30 : 0.4393 0.4407  !
Topics 151-200: ql+rm3   : p30 : 0.4987 0.4993  !

# Robust04
All Topics: bm25+rm3 : map : 0.2726 0.2757  !
All Topics: ql+rm3   : map : 0.2615 0.2645  !
All Topics: bm25+rm3 : p30 : 0.3237 0.3256  !
All Topics: ql+rm3   : p30 : 0.3154 0.3153  !

# Robust05
TREC 2005 Robust Track Topics: bm25+rm3 : map : 0.2448 0.2511  !
TREC 2005 Robust Track Topics: ql+rm3   : map : 0.2422 0.2480  !
TREC 2005 Robust Track Topics: bm25+rm3 : p30 : 0.3913 0.3873  !
TREC 2005 Robust Track Topics: ql+rm3   : p30 : 0.3933 0.4007  !

# Core17
All Topics: bm25+rm3 : map : 0.2534 0.2633  !
All Topics: ql+rm3   : map : 0.2363 0.2409  !
All Topics: bm25+rm3 : p30 : 0.4767 0.4893  !
All Topics: ql+rm3   : p30 : 0.4640 0.4647  !

# Wt10g
Wt10g: Topics 451-550: bm25+rm3 : map : 0.2187 0.2166  !
Wt10g: Topics 451-550: ql+rm3   : map : 0.2181 0.2173  !
Wt10g: Topics 451-550: bm25+rm3 : p30 : 0.2463 0.2452  !
Wt10g: Topics 451-550: ql+rm3   : p30 : 0.2316 0.2344  !

# Gov2
Topics 701-750: bm25+rm3 : map : 0.2952 0.2981  !
Topics 751-800: bm25+rm3 : map : 0.3840 0.3843  !
Topics 801-850: bm25+rm3 : map : 0.3410 0.3439  !
Topics 701-750: ql+rm3   : map : 0.2799 0.2777  !
Topics 751-800: ql+rm3   : map : 0.3628 0.3620  !
Topics 801-850: ql+rm3   : map : 0.3199 0.3168  !
Topics 701-750: bm25+rm3 : p30 : 0.5299 0.5333  !
Topics 751-800: bm25+rm3 : p30 : 0.5933 0.5933
Topics 801-850: bm25+rm3 : p30 : 0.5260 0.5240  !
Topics 701-750: ql+rm3   : p30 : 0.4850 0.4912  !
Topics 751-800: ql+rm3   : p30 : 0.5673 0.5700  !
Topics 801-850: ql+rm3   : p30 : 0.4880 0.4840  !

# ClueWeb09b
Topics  51-100: bm25+rm3 : map  : 0.1066 0.1077  !
Topics 101-150: bm25+rm3 : map  : 0.1142 0.1147  !
Topics 151-200: bm25+rm3 : map  : 0.1335 0.1319  !
Topics  51-100: ql+rm3   : map  : 0.1055 0.1061  !
Topics 101-150: ql+rm3   : map  : 0.1021 0.0963  !
Topics 151-200: ql+rm3   : map  : 0.1120 0.1140  !
Topics  51-100: bm25+rm3 : p30  : 0.2583 0.2556  !
Topics 101-150: bm25+rm3 : p30  : 0.2627 0.2687  !
Topics 151-200: bm25+rm3 : p30  : 0.2313 0.2400  !
Topics  51-100: ql+rm3   : p30  : 0.2514 0.2535  !
Topics 101-150: ql+rm3   : p30  : 0.2267 0.2187  !
Topics 151-200: ql+rm3   : p30  : 0.2007 0.2053  !
Topics  51-100: bm25+rm3 : ndcg : 0.1439 0.1430  !
Topics 101-150: bm25+rm3 : ndcg : 0.1921 0.1885  !
Topics 151-200: bm25+rm3 : ndcg : 0.1309 0.1254  !
Topics  51-100: ql+rm3   : ndcg : 0.1305 0.1328  !
Topics 101-150: ql+rm3   : ndcg : 0.1670 0.1608  !
Topics 151-200: ql+rm3   : ndcg : 0.1027 0.1084  !

# ClueWeb12-B13
Topics 701-750: bm25+rm3 : map  : 0.0426 0.0438  !
Topics 751-800: bm25+rm3 : map  : 0.0188 0.0188
Topics 701-750: ql+rm3   : map  : 0.0320 0.0314  !
Topics 751-800: ql+rm3   : map  : 0.0201 0.0204  !
Topics 701-750: bm25+rm3 : p30  : 0.1753 0.1773  !
Topics 751-800: bm25+rm3 : p30  : 0.1133 0.1140  !
Topics 701-750: ql+rm3   : p30  : 0.1447 0.1433  !
Topics 751-800: ql+rm3   : p30  : 0.1173 0.1160  !
Topics 701-750: bm25+rm3 : ndcg : 0.1116 0.1172  !
Topics 751-800: bm25+rm3 : ndcg : 0.0991 0.0987  !
Topics 701-750: ql+rm3   : ndcg : 0.0876 0.0878  !
Topics 751-800: ql+rm3   : ndcg : 0.0964 0.0961  !

# ClueWeb12
Topics 701-750: bm25+rm3 : map  : 0.1501 0.1483  !
Topics 751-800: bm25+rm3 : map  : 0.2434 0.2460  !
Topics 701-750: ql+rm3   : map  : 0.1205 0.1245  !
Topics 751-800: ql+rm3   : map  : 0.2321 0.2302  !
Topics 701-750: bm25+rm3 : p30  : 0.2333 0.2360  !
Topics 751-800: bm25+rm3 : p30  : 0.4113 0.4167  !
Topics 701-750: ql+rm3   : p30  : 0.2040 0.2053  !
Topics 751-800: ql+rm3   : p30  : 0.3940 0.3880  !
Topics 701-750: bm25+rm3 : ndcg : 0.1693 0.1752  !
Topics 751-800: bm25+rm3 : ndcg : 0.2412 0.2479  !
Topics 701-750: ql+rm3   : ndcg : 0.1503 0.1563  !
Topics 751-800: ql+rm3   : ndcg : 0.2183 0.2139  !

# Tweets2011
TREC 2011 Microblog Track: bm25+rm3 : map : 0.3467 0.3472  !
TREC 2012 Microblog Track: bm25+rm3 : map : 0.2057 0.2056  !
TREC 2011 Microblog Track: ql+rm3   : map : 0.4102 0.4100  !
TREC 2012 Microblog Track: ql+rm3   : map : 0.2413 0.2413
TREC 2011 Microblog Track: bm25+rm3 : p30 : 0.4041 0.4027  !
TREC 2012 Microblog Track: bm25+rm3 : p30 : 0.3429 0.3418  !
TREC 2011 Microblog Track: ql+rm3   : p30 : 0.4476 0.4463  !
TREC 2012 Microblog Track: ql+rm3   : p30 : 0.3537 0.3531  !

# Tweets2013
TREC 2013 Microblog Track: bm25+rm3 : map : 0.2359 0.2356  !
TREC 2014 Microblog Track: bm25+rm3 : map : 0.3964 0.3962  !
TREC 2013 Microblog Track: ql+rm3   : map : 0.2803 0.2804  !
TREC 2014 Microblog Track: ql+rm3   : map : 0.4660 0.4660
TREC 2013 Microblog Track: bm25+rm3 : p30 : 0.4056 0.4061  !
TREC 2014 Microblog Track: bm25+rm3 : p30 : 0.5873 0.5873
TREC 2013 Microblog Track: ql+rm3   : p30 : 0.4617 0.4606  !
TREC 2014 Microblog Track: ql+rm3   : p30 : 0.6467 0.6442  !
```

### CODE REVIEW

**Comment by lintool** [comment]
@Peilin-Yang Addressed CR, renamed `ExtractRm3Stopwords` to `ExtractTopDfTerms`
